Detectors/FIT/FT0/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,189 @@
+// or submit itself to any jurisdiction.
+/*
+  Digits to RAW data coding. RAW data format - class DataFormat/Detectors/FIT/FT0/RawEventData
+  18 PMs (GBT links) 12 MCPs each  and 1 TCM, each stream transmit separately
+Event header - 80bits
+  uint startDescriptor : 4;
+  uint nGBTWords : 4;
+  uint reservedField : 32;
+  uint orbit : 32;
+  uint bc : 12;
+
+  Event data 40bits
+  short int time : 12;
+  short int charge : 12;
+  unsigned short int numberADC : 1;
+  bool isDoubleEvent : 1;
+  bool is1TimeLostEvent : 1;
+  bool is2TimeLostEvent : 1;
+  bool isADCinGate : 1;
+  bool isTimeInfoLate : 1;
+  bool isAmpHigh : 1;
+  bool isEventInTVDC : 1;
+  bool isTimeInfoLost : 1;
+  uint channelID : 4;
+GBT packet:
+RDH + Event header + event data, 2 channels per 1 GBT word;
+if no data for this PM - only headers.
+
+Trigger mode : detector sends data to FLP at each trigger;
+Continueous mode  :   for only bunches with data at least in 1 channel.  
+*/
+
+#include "Headers/RAWDataHeader.h"
+#include "CommonDataFormat/InteractionRecord.h"
+#include "DataFormatsFT0/RawEventData.h"
+#include "FT0Simulation/Digits2Raw.h"
+#include "DetectorsBase/Triggers.h"
+#include <Framework/Logger.h>
+#include <TStopwatch.h>
+#include <cassert>
+#include <fstream>
+#include <vector>
+#include <bitset>
+#include <iomanip>
+#include "TFile.h"
+#include "TTree.h"
+
+using namespace o2::ft0;
+
+ClassImp(Digits2Raw);
+void setRDH(o2::header::RAWDataHeader&, int nlink, o2::InteractionRecord const& mIntRecord, int nPages);
+EventHeader makeGBTHeader(int link, o2::InteractionRecord const& mIntRecord);
+
+Digits2Raw::Digits2Raw(const char* fileRaw, std::string fileDigitsName)
+{
+
+  mFileDest.exceptions(std::ios_base::failbit | std::ios_base::badbit);
+  mFileDest.open(fileRaw, std::fstream::out | std::fstream::binary);
+  Digits2Raw::readDigits(fileDigitsName.c_str());
+}
+
+void Digits2Raw::readDigits(const char* fileDigitsName)
+{
+  LOG(INFO) << "**********Digits2Raw::convertDigits" << std::endl;
+
+  o2::ft0::LookUpTable lut{o2::ft0::Digits2Raw::linear()};
+  LOG(INFO) << " ##### LookUp set ";
+
+  TFile* fdig = TFile::Open(fileDigitsName);
+  assert(fdig != nullptr);
+  LOG(INFO) << " Open digits file ";
+  TTree* digTree = (TTree*)fdig->Get("o2sim");
+  std::vector<o2::ft0::Digit>* digArr = new std::vector<o2::ft0::Digit>;
+  digTree->SetBranchAddress("FT0Digit", &digArr);
+  Int_t nevD = digTree->GetEntries(); // digits in cont. readout may be grouped as few events per entry
+  uint32_t old_orbit = 0;
+  for (Int_t iev = 0; iev < nevD; iev++) {
+    digTree->GetEvent(iev);
+    for (const auto& digit : *digArr) {
+      auto mIntRecord = digit.getInteractionRecord();
+      uint32_t current_orbit = mIntRecord.orbit;
+      if (old_orbit != current_orbit) {
+        for (DataPageWriter& writer : mPages)
+          writer.flush(mFileDest, mNpages);
+        old_orbit = current_orbit;
+        mNpages = 0;
+      }
+      convertDigits(digit, lut);
+    }
+  }
```

Comment:
should not you also flush the data filled in the last iteration of the convertDigits?